### PR TITLE
Promote: ConfigMap Lifecycle test - +2 conformance endpoint coverage

### LINKED ISSUE #90921: Write CoreV1ConfigMap test - +2 endpoint Coverage
# Progress `[6/7]`
* [x] Original approval issue: #87120
* [x]  Test approval issue: [kubernetes/kubernetes#90921](https://github.com/kubernetes/kubernetes/issues/90921)
* [x]  Test pr: [PR](https://github.com/kubernetes/kubernetes/pull/87212)
* [x]  Two weeks soak start date: 24 April 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=should%20run%20through%20a%20ConfigMap%20lifecycle)
* [x]  Two weeks soak end date: 8 May 2020
* [ ]  Promote pr: [With Watch tool](https://github.com/kubernetes/kubernetes/pull/90941) 
* * [x]  [update to improve](https://github.com/kubernetes/kubernetes/pull/90944)
* [x]  Two weeks soak start date: 8 June 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=should%20run%20through%20a%20ConfigMap%20lifecycle)
* [x]  Two weeks soak end date: 23 June 2020
This issue is created to allow edit by @Riaankl

/wip
/hold

# Notes
This issue proposes the writing of a new test for the API endpoints `listCoreV1ConfigMapForAllNamespaces`, `listCoreV1ConfigMapForAllNamespaces` with the intention of promoting it to Conformance. This issue contains a **very basic mock test** in order to start the conversation about what a good e2e test would look like for this endpoint. The code below is verified to be hitting the intended endpoint, show as per the queries in APIsnoop's live view of the cluster's audit logs.

## Deploy APISnoop
```shell
kubectl apply -f ~/cncf/apisnoop/deployment/k8s/raiinbow.yaml
```

# Identify an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining ConfigMap endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  description
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and operation_id ilike '%ConfigMap%'
 ORDER BY operation_id desc
 LIMIT 150
       ;
```

```
               operation_id                |               description               
-------------------------------------------|-----------------------------------------
 listCoreV1ConfigMapForAllNamespaces       | list or watch objects of kind ConfigMap
 deleteCoreV1CollectionNamespacedConfigMap | delete collection of ConfigMap
(2 rows)
```

# Use API Reference to Lightly Document the Feature
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - ConfigMap](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/configmap.go)

# The mock test
## Test outline
1. Create a ConfigMap with a static label
2. Patch the ConfigMap with a new Label and updated data
3. Get the ConfigMap to ensure it's patched
4. List all ConfigMaps in all Namespaces find the ConfigMap(1) ensure that the ConfigMap is found and is patched
5. Delete Namespaced ConfigMap(1) via a Collection with a LabelSelector

## Example in Go
```shell
go get -v -u k8s.io/apimachinery/pkg/apis/meta/v1
go get -v -u k8s.io/client-go/kubernetes
go get -v -u k8s.io/client-go/tools/clientcmd
```

This mock test uses client-go's standard functions for talking to the Event area of the API.

```go
package main

import (
  "fmt"
  "flag"
  "os"
  "encoding/json"
  v1 "k8s.io/api/core/v1"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
    fmt.Println(err)
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)

  testNamespaceName := "default"
  testConfigMapName := "test-configmap"

  _, err = ClientSet.CoreV1().ConfigMaps(testNamespaceName).Create(&v1.ConfigMap{
    ObjectMeta: metav1.ObjectMeta{
      Name: testConfigMapName,
      Labels: map[string]string{
        "test-configmap-static": "true",
      },
    },
    Data: map[string]string{
      "valueName": "value",
    },
  })
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] created ConfigMap")

  configMapPatchPayload, err := json.Marshal(v1.ConfigMap{
    ObjectMeta: metav1.ObjectMeta{
      Labels: map[string]string{
        "test-configmap": "patched",
      },
    },
    Data: map[string]string{
      "valueName": "value1",
    },
  })
  if err != nil {
    fmt.Println(err)
    return
  }
  _, err = ClientSet.CoreV1().ConfigMaps(testNamespaceName).Patch(testConfigMapName, types.StrategicMergePatchType, []byte(configMapPatchPayload))
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] patched ConfigMap")

  configMap, err := ClientSet.CoreV1().ConfigMaps(testNamespaceName).Get(testConfigMapName, metav1.GetOptions{})
  if err != nil {
    fmt.Println(err)
    return
  }
  if configMap.Data["valueName"] != "value1" || configMap.Labels["test-configmap"] != "patched" {
    fmt.Println("[error] failed to patch ConfigMap")
    return
  }
  fmt.Println("[status] ConfigMap was patched successully")

  // listing in all namespaces to hit the endpoint
  configMapList, err := ClientSet.CoreV1().ConfigMaps("").List(metav1.ListOptions{
    LabelSelector: "test-configmap-static=true",
  })
  if err != nil {
    fmt.Println(err)
    return
  }
  if len(configMapList.Items) == 0 {
    fmt.Println("[error] no ConfigMaps found")
    return
  }
  testConfigMapFound := false
  for _, cm := range configMapList.Items {
    if cm.ObjectMeta.Name == testConfigMapName &&
       cm.ObjectMeta.Namespace == testNamespaceName &&
       cm.ObjectMeta.Labels["test-configmap-static"] == "true" &&
       cm.Data["valueName"] == "value1" {
      testConfigMapFound = true
      break
    }
  }
  if testConfigMapFound == false {
    fmt.Println("[error] unable to find ConfigMap in list")
    return
  }
  fmt.Println("[status] found updated ConfigMap")

  err = ClientSet.CoreV1().ConfigMaps(testNamespaceName).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{
    LabelSelector: "test-configmap-static=true",
  })
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] deleted ConfigMap")

  fmt.Println("[status] complete")

}
```

# Verify with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

Display endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |               operation_id                | hit_by_ete | hit_by_new_test 
-------------------|-------------------------------------------|------------|-----------------
 live-test-writing | createCoreV1NamespacedConfigMap           |        288 |               2
 live-test-writing | deleteCoreV1CollectionNamespacedConfigMap |          0 |               2
 live-test-writing | listCoreV1ConfigMapForAllNamespaces       |          0 |               1
 live-test-writing | patchCoreV1NamespacedConfigMap            |          2 |               2
 live-test-writing | readCoreV1NamespacedConfigMap             |         23 |               1
(5 rows)
```

Show the change in coverage after the mock test:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------|-----------------|--------------|--------------|------------------
 test_coverage |             438 |          183 |          185 |                2
(1 row)
```

# Final notes
If a test with these calls gets merged, ****Conformance coverage will go up by 2 points****

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Promotes the following E2E tests to Conformance:
`test/e2e/common/configmap.go: "should run through a ConfigMap lifecycle"`

**Special notes for your reviewer**:
Adds +2 conformance endpoint coverage.
Testgrid: [results](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=should%20run%20through%20a%20ConfigMap%20lifecycle)
Fixes: #90921

**Does this PR introduce a user-facing change?**:

```
NONE
```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```
NONE
```

/area conformance
/area test
@kubernetes/sig-architecture-pr-reviews
@kubernetes/cncf-conformance-wg

### CODE REVIEW

**Comment by Riaankl** [comment]
Comments from #88034
https://github.com/kubernetes/kubernetes/pull/88034#issuecomment-610085217
Feedback requested from #k8s-conformance slack channel: https://kubernetes.slack.com/archives/C78F00H99/p1587417924039800

**Comment by Riaankl** [comment]
This PR replace #88034

**Comment by BobyMCbobs** [comment]
Rebased to include the changes from [Update and improve ConfigMap resource lifecycle test](https://github.com/kubernetes/kubernetes/pull/90944) which depends on [Add watch event tooling
](https://github.com/kubernetes/kubernetes/pull/91416)

**Comment by Riaankl** [comment]
@johnbelamaric @spiffxp 
Can we please get an LGTM and Approve

**Comment by Riaankl** [comment]
@johnbelamaric @spiffxp
Can we please get an LGTM and Approve

**Comment by spiffxp** [comment]
/hold
Not yet, for a few reasons:
- The test needs 2 weeks soak time before it can be promoted, https://github.com/kubernetes/kubernetes/pull/90944 merged 3 days ago
- This PR has many more changes in it than just promoting, it needs to be rebased
- I'm still not sold on the extra complexity in this test being worth it

**Comment by Riaankl** [comment]
@BobyMCbobs 

Please review:
> Not yet, for a few reasons:
> 
> * The test needs 2 weeks soak time before it can be promoted, #90944 merged 3 days ago
> * This PR has many more changes in it than just promoting, it needs to be rebased
> * I'm still not sold on the extra complexity in this test being worth it

**Comment by BobyMCbobs** [comment]
> /hold
> Not yet, for a few reasons:
> 
> * The test needs 2 weeks soak time before it can be promoted, #90944 merged 3 days ago
> * This PR has many more changes in it than just promoting, it needs to be rebased
> * I'm still not sold on the extra complexity in this test being worth it

@spiffxp, I've rebased. Tomorrow it will be soaked for two weeks.

Our simple approach was to ensure that given watch events occur using the _WatchEventSequenceVerifier_.
For each API call, using _WatchUntilWithoutRetry_ to ensure that specific watch events occurred and/or that the resource has certain values at given points in the test. This is repeated 3 times with Added, Modified, and Deleted. As we iterate at over this pattern, we may see use for a library - for now, we want to make sure that we discover a consistent pattern.

**Comment by spiffxp** [comment]
I opened https://github.com/kubernetes/kubernetes/pull/92621 as a suggestion of how I would simplify this test to avoid watches while still ensuring create/update/delete take effect

**Comment by spiffxp** [comment]
I will grant that it looks ok from a testgrid perspective: https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=should%20run%20through%20a%20ConfigMap%20lifecycle

So if we feel compelled to promote it as is, fine.  But I still think that this shouldn't be an example we follow for how to write e2e tests

**Comment by BobyMCbobs** [comment]
It's passing and has been sitting here for some time.
We can updated it at a later date.
Can this please be approved?

**Comment by Riaankl** [comment]
@johnbelamaric 
Can we please get a /approve and /LGTM?

**Comment by Riaankl** [comment]
@smarterclayton Prowbot is also looking for your /approve

**Review by oomichi — COMMENTED** [review]
The test itself seems stable as https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-gce&include-filter-by-regex=should%20run%20through%20a%20ConfigMap%20lifecycle

It is fine to promote it to conformance for me.
just one comment for format thing.

File: test/e2e/common/configmap.go
Comment (oomichi) [reply test/e2e/common/configmap.go:172]: /cc @oomichi

Comment (heyste) [reply test/e2e/common/configmap.go]: Formatting fixed.

Comment (Riaankl) [reply test/e2e/common/configmap.go]: > The indent seems a little odd by comparing another conformance tests like line 142.
@oomichi 
@johnbelamaric 
Code indent fixed. Can we please get /LGTM & /approve ?

**Comment by spiffxp** [comment]
/milestone v1.19
SIG Testing chair, test improvement ahead of test freeze

**Comment by oomichi** [comment]
Thanks for updating.

/lgtm

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Pull request merged**
